enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,424 @@
+---
+title: machine-deletion-hooks
+authors:
+  - "@JoelSpeed"
+  - "@michaelgugino"
+reviewers:
+  - "@enxebre"
+  - "@elmiko"
+  - "@ademicev"
+  - "@hexfusion"
+approvers:
+  - TBD
+creation-date: 2021-08-10
+last-updated: 2021-08-10
+status: implementable
+see-also:
+  - [Cluster API Upstream Equivalent](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20200602-machine-deletion-phase-hooks.md)
+---
+
+# Machine Deletion Hooks
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Glossary
+
+### Lifecycle Hook
+A specific point in a machine's reconciliation lifecycle where execution of
+normal machine-controller behaviour is paused or modified.
+
+### Deletion Phase
+Describes when a machine has been marked for deletion but is still present
+in the API.  Various actions happen during this phase, such as draining a node,
+deleting an instance from a cloud provider, and deleting the node object.
+
+### Hook Implementing Controller (HIC)
+The Hook Implementing Controller describes a controller, other than the
+machine controller, that adds, removes, and/or responds to a particular
+lifecycle hook. Each lifecycle hook should have a single HIC, but an HIC
+can optionally manage one or more hooks.
+
+## Summary
+
+Defines a set of annotations that can be applied to a Machine which affect the
+linear progress of a Machine’s lifecycle after a Machine has been marked for
+deletion. These annotations are optional and may be applied during Machine
```

Comment:
can they be modified after deletion and before api removal?

Reply:
> Yes, they need to be modified because they have to be removed before the Machine can be deleted. In theory new ones could be added, I don't think this really affects us too much as we expect these will always be added before the machine is deleted, if new ones are added post delete, we will block again, but the result may not be what is intended (eg we already called for the cloud provider to terminate the instance and then someone adds a pre-drain).

I suggest getting specific here.  It seems reasonable to prevent the addition of new hooks after deletion has begun.